File: behaviorial/iterator/python/concrete_iterator.py
from interface import Iterator


class TreeNode:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right


class DFSIterator(Iterator):
    def __init__(self, root: TreeNode):
        self.stack = []
        if root:
            self.stack.append(root)

    def has_next(self) -> bool:
        return len(self.stack) > 0

    def next(self):
        if not self.has_next():
            return None
        node = self.stack.pop()
        if node.right:
            self.stack.append(node.right)
        if node.left:
            self.stack.append(node.left)
        return node.value

File: behaviorial/iterator/typescript/concrete-iterator.ts
import { TreeNode } from './concrete-collection';
import { Iterator } from './interface';

export class DFSIterator<T> implements Iterator<T> {
	private stack: (TreeNode<T> | null)[] = [];

	constructor(root: TreeNode<T> | null) {
		if (root) {
			this.stack.push(root);
		}
	}

	hasNext(): boolean {
		return this.stack.length > 0;
	}

	next(): T | null {
		if (!this.hasNext()) {
			return null;
		}
		const node = this.stack.pop();
		if (node?.right) {
			this.stack.push(node.right);
		}
		if (node?.left) {
			this.stack.push(node.left);
		}
		return node?.value ?? null;
	}
}
